### DISCUSSION TITLE
While running with splunk-otel-agent 1.2, can API 1.4.1 create custom spans via Tracer?

### DISCUSSION THREAD

**Discussion by woldie** [signalfx/splunk-otel-java#431] [Q&A]
I am auto-instrumenting with the splunk-otel-agent 1.2, and I'd like to create custom spans using the same Tracer as the one the agent uses. I am currently attempting to acquire the Tracer with API 1.4.1 running on OpenJDK 11:

```
GlobalOpenTelemetry.getTracer(Objects.requireNonNull(applicationContext.getId()))
```

And I am trying to create custom spans with

```
tracer.spanBuilder("mySpan")
        .setSpanKind(SpanKind.SERVER)
        .startSpan();
```

(I have also tried SpanKind.INTERNAL and left setSpanKind() off of the expression, but it made no difference)

Unfortunately, my custom spans are not traced, and only the auto-instrumented spans the agent creates show up in my traces in SignalFX. I am calling Span#end on all my custom spans.

This link referenced in another similar answer is no longer relevant: https://github.com/open-telemetry/opentelemetry-java/blob/main/QUICKSTART.md#tracing

So, I would like to know the following given the latest and greatest splunk-otel-agent:

1. Is it possible to access/connect to the same Tracer that splunk-otel-agent must have internally?
2. If yes, is it possible (without using @WithSpan) to add custom spans to my service via that Tracer?

Thanks!

EDIT 1:  I see two conflicting comments in the splunk-otel documentation.  At this link, https://docs.splunk.com/Observability/gdi/get-data-in/application/java/instrumentation/instrument-java-application.html there is a note that reads:

```
If you need to add custom attributes to spans or want to manually generate spans, you need to manually
instrument your Java application or service. See Manually instrument Java applications.
```

but on the "Manually instrument Java applications" page, a note reads:

```
Manual OTel instrumentation is fully compatible with Splunk automatic JVM instrumentation and is fully
supported by Splunk.
```

So which is it?

**Comment by mateuszrzeszutek** [comment]
Hey @woldie ,
Can you share some more details about how you're using the OpenTelemetry API in your app? A small repo containing a reproduction scenario would be invaluable 🙏 

> This link referenced in another similar answer is no longer relevant: https://github.com/open-telemetry/opentelemetry-java/blob/main/QUICKSTART.md#tracing

It points to the official [OpenTelemetry Java docs](https://opentelemetry.io/docs/java/manual_instrumentation/) page that described how to use OpenTelemetry API in detail.

> 1. Is it possible to access/connect to the same Tracer that splunk-otel-agent must have internally?
> 2. If yes, is it possible (without using @WithSpan) to add custom spans to my service via that Tracer?

1. Internally the agent uses many different `Tracer` instances - but you definitely should be able to add your own telemetry to the agent's output.
2. Yes, most definitely - the scenario that you've described above should work.

> EDIT 1: I see two conflicting comments in the splunk-otel documentation. At this link, https://docs.splunk.com/Observability/gdi/get-data-in/application/java/instrumentation/instrument-java-application.html there is a note that reads:
> 
> ```
> If you need to add custom attributes to spans or want to manually generate spans, you need to manually
> instrument your Java application or service. See Manually instrument Java applications.
> ```
> 
> but on the "Manually instrument Java applications" page, a note reads:
> 
> ```
> Manual OTel instrumentation is fully compatible with Splunk automatic JVM instrumentation and is fully
> supported by Splunk.
> ```

They're both correct: if you want to add custom attributes or spans, you need to manually instrument your application - AND manual instrumentation is fully supported by Splunk.

**Reply by woldie** [reply]
> They're both correct: if you want to add custom attributes or spans, you need to manually instrument your application - AND manual instrumentation is fully supported by Splunk.

Ok, so I can't manually add spans AND run the agent?  I take this to mean that I need to manually instrument everything and miss out on all the agent's integrations if I want to do custom spans.

Are you confirming that?  If so, that's kindof a bummer and makes the agent a limited item that doesn't realize the OTEL API.  Do you have a roadmap on when/if the agent will be able to pick up custom spans?  I'd need to add custom BatchSpanProcessors to the agent's Tracer as well as long as we're getting into feature requests.

> > This link referenced in another similar answer is no longer relevant: https://github.com/open-telemetry/opentelemetry-java/blob/main/QUICKSTART.md#tracing

> It points to the official OpenTelemetry Java docs page that described how to use OpenTelemetry API in detail.

That #tracing anchor tag on the link I referenced used to discuss custom spans with the agent.  Recent versions of that page have all that content stripped.

> Can you share some more details about how you're using the OpenTelemetry API in your app? A small repo containing a reproduction scenario would be invaluable 🙏

Let me hear from you whether the agent is a non-starter for custom spans or not and if it is then I can try to rig a repo with a code sample.  I am in no hurry to adopt splunk-otel as I have an in-house tracing solution that works.  I can change my custom span support code to adapt to whatever needs to happen in the OTEL api's.

**Reply by jkwatson** [reply]
> > They're both correct: if you want to add custom attributes or spans, you need to manually instrument your application - AND manual instrumentation is fully supported by Splunk.
> 
> Ok, so I can't manually add spans AND run the agent? I take this to mean that I need to manually instrument everything and miss out on all the agent's integrations if I want to do custom spans.
> 
> Are you confirming that? If so, that's kindof a bummer and makes the agent a limited item that doesn't realize the OTEL API. Do you have a roadmap on when/if the agent will be able to pick up custom spans? I'd need to add custom BatchSpanProcessors to the agent's Tracer as well as long as we're getting into feature requests.

No! You can *definitely* use the agent and do manually instrumentation that lives alongside it. This is definitely a fully supported and expected mode of operation.


> > > This link referenced in another similar answer is no longer relevant: https://github.com/open-telemetry/opentelemetry-java/blob/main/QUICKSTART.md#tracing
> 
> > It points to the official OpenTelemetry Java docs page that described how to use OpenTelemetry API in detail.
> 
> That #tracing anchor tag on the link I referenced used to discuss custom spans with the agent. Recent versions of that page have all that content stripped.

That documentation is, I think, living in the opentelemetry-java-instrumentation repository now, but it needs to get pushed out to the website, for sure.

**Reply by woldie** [reply]
> No! You can definitely use the agent and do manually instrumentation that lives alongside it. This is definitely a fully supported and expected mode of operation.

Ok, that's great news.  I'm going to assume "alongside it" means my custom spans can use the same configured Tracer as the one configured by the agent.

In talking with others in my org, it seems like running the agent and making custom spans is supported and they are having success with it, suggesting that I'm doing something wrong like starting the Span but not makeCurrent()'ing it?  I'll get to the bottom of what I'm doing different from them and report back here if I see success.

**Reply by jkwatson** [reply]
Btw, I think you want to say the "same configured SDK", not Tracer. there are very many Tracers in a given SDK, and you just need to make sure you're using one from the same SDK.

**Reply by woldie** [reply]
Thanks @jkwatson, good distinction.

I am happy to report that I was able to get my agent-driven app to send custom spans and they appeared in SignalFX.  I found that I needed to call Span.makeCurrent() to connect my started spans to the current parent span.

Also, I found I had some of my final custom spans associated with the traceId were not making it to SignalFX.  But, when I did a span.setStatus(StatusCode.OK) just prior to scope.close() and span.end(), all of my custom spans made it to SignalFX.  Not sure if I was witnessing a race condition in the agent/SDK or if there is more pilot error in my use of the system that the setStatus(OK) allowed me to work around.

In any case, thanks all for your help today.  I think we've proven that custom spans will work.

Next thing to discover is whether we can add a custom BatchSpanProcessor to the same configured SDK as the agent's.  (we want to do some custom logging at the end of every span, regardless of whether it is a custom span or agent-created span.)

**Reply by mateuszrzeszutek** [reply]
> Next thing to discover is whether we can add a custom BatchSpanProcessor to the same configured SDK as the agent's. (we want to do some custom logging at the end of every span, regardless of whether it is a custom span or agent-created span.)

If it's just some logging, please take a look at the [`SpanProcessor`](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/SpanProcessor.java) interface - it's definitely more lightweight than running another `BatchSpanProcessor`.

**Reply by jkwatson** [reply]
> Also, I found I had some of my final custom spans associated with the traceId were not making it to SignalFX. But, when I did a span.setStatus(StatusCode.OK) just prior to scope.close() and span.end(), all of my custom spans made it to SignalFX. Not sure if I was witnessing a race condition in the agent/SDK or if there is more pilot error in my use of the system that the setStatus(OK) allowed me to work around.

This doesn't make any sense to me. Setting a span status should not change this at all. Can you enable debug logging in the agent and make sure that the spans are getting created, whether or not the the status has been explicitly set?

**Reply by woldie** [reply]
@mateuszrzeszutek thanks for the tip, that looks like exactly what I want

**Reply by woldie** [reply]
This page describes how to extend the agent with a custom SpanProcessor:  https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/examples/extension/README.md

**Reply by mateuszrzeszutek** [reply]
> This page describes how to extend the agent with a custom SpanProcessor: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/examples/extension/README.md

Yes, that's correct.

**Comment by furkanali0325** [comment]
Hi Team,
I have integrated Splunk OTEL jar( open telemetry library) in my spring boot application for tracing and matrices. 
   So there is a SPLUNK_ACCESS_TOKEN that needs to be exposed to send data directly to the Splunk observability cloud. Currently, I have stored it in Kubernetes.yaml file. So I just want to know other approaches to expose this Splunk access token.

For example, there is a flag variable "OTEL_JAVAAGENT_ENABLED" used by the Splunk OTEL library and I have stored this value to "true" in Kubernetes.yaml file. So by this config OTEL jar is able to access it.

**Is there any alternate way to store the splunk_access_token apart from Kubernetes.yaml or properties file?**
